### DISCUSSION TITLE
No rules are being processed

### DISCUSSION THREAD

**Discussion by Crowiiie** [Azure/PSRule.Rules.Azure#1426] [Q&A] [answered]
Hi,

I have already checked out the https://github.com/Azure/PSRule.Rules.Azure/discussions/1332, but unfortunately keep running into a wall with getting this to work.

I have set up a _.ps-rule_ folder within my repository, which contains the files bicep.Rule.ps1:
![image](https://user-images.githubusercontent.com/62597625/168993205-6441bb97-8c58-4aea-82a4-7d4783298cd9.png)

I have also set up a _templates_ folder, which contains a main.bicep file:
![image](https://user-images.githubusercontent.com/62597625/168993347-7ab00538-9377-410a-b129-0749fc0445e9.png)

I have set up the ps-rule.yaml file in the root:
![image](https://user-images.githubusercontent.com/62597625/168993632-e9528225-517d-428a-9ba4-5e40cfa24eee.png)

I then have a pipeline which installs the dependencies and runs the Assert-PSRule:
![image](https://user-images.githubusercontent.com/62597625/168993529-a7eb92d8-74c8-46a8-a2c6-17aa716dcaf7.png)

Unfortunately it seems like it is unable to find my custom rule, given that it processes 0 rules:
![image](https://user-images.githubusercontent.com/62597625/168994246-bc5d8854-4bf2-436d-b1da-92c091f18473.png)

I assume I am missing something somewhere, but I have unfortunately gotten to a point where I don't know where to look anymore. I have attempted using the example in the documentation (https://microsoft.github.io/PSRule/v2/scenarios/validation-pipeline/validation-pipeline/#complete-example), but it only works when the folder to analyze is the .ps-rule. This seems like an awesome extension and I would really like to get it working. 

Any idea where I am going wrong? :)

**Comment by BernieWhite** [comment, marked as answer]
@Crowiiie Thanks for the question. Three things to try that should resolve this.

## Use PSRule for Azure

Azures resources from a Bicep file are a little tricky because they need to be expanded (to resolve parameters, variables, functions, and modules). We've already done a lot of the work through PSRule for Azure. Have a look here [Installation](https://azure.github.io/PSRule.Rules.Azure/install-instructions/#withazurepipelines) and [Creating your pipeline](https://azure.github.io/PSRule.Rules.Azure/creating-your-pipeline/). The `AZURE_BICEP_FILE_EXPANSION` configuration option in your `ps-rule.yaml` is for PSRule for Azure, and won't work if it is not installed.

If you do not want to use the Azure Pipeline task installable from the Visual Studio Marketplace suggested by these docs, which is the simplest path, then:

- Install PSRule for Azure by using `Install-Module PSRule.Rules.Azure`, this would also install the `PSRule` module automatically, so a separate command is not required.
- Update your `Assert-PSRule` command-line to include `-Module PSRule.Rules.Azure -Format File` to use the PSRule for Azure module.
- Set `-InputPath` to where you bicep files are or the whole repo instead of `.ps-rule/`. From whichever path you choose, PSRule will automatically scan all subfolders. For example `-InputPath .` or `-InputPath '**/*.bicep'`.

## Update the type for the custom rule

Instead of using `Microsoft.Storage/storageAccounts@2021-09-01` in your rule, just use `Microsoft.Storage/storageAccounts`. The `@2021-09-01` refers to the API version which is not strictly part of the resource type.

## Configure type bindings for local rules

Set the `binding.targetType` option in `ps-rule.yaml` as described here [Binding type](https://azure.github.io/PSRule.Rules.Azure/customization/enforce-custom-tags/#bindingtype).

For example:

```yaml
# Configure binding options
binding:
  targetType:
  - 'resourceType'
  - 'type'
```

---

I hope that helps.

**Comment by Crowiiie** [comment]
Thank you for taking the time to write out such a comprehensive answer! I have implemented the changes suggested above, and now both the Azure Pipeline task and the Powershell script options are working properly. 👍
